test/known_problems/should_pass/inner_union_subtype_of_root_union.erl

```diff
@@ -0,0 +1,18 @@
+-module(inner_union_subtype_of_root_union).
+
+-export([f/0]).
+
+% The problem is that {t, a} | {t, b} is not subtype of {t, a|b}.
+
+% It has surfaced because
+%     {'type', anno(), 'map_field_assoc' | 'map_field_exact', [abstract_type()]}
+% is not a subtype of
+%     {'type', anno(), 'map_field_assoc', [abstract_type()]}
+%   | {'type', anno(), 'map_field_exact', [abstract_type()]}
```

Comment:
Could we try to solve this by `normalize`ing them to the same structure?

Reply:
Hmm, maybe we could. But I'm not sure what should be criterion/algorithm for merging. We probably agree that `{t, a} | {t, b}` should be normalized into `{t, a|b}`. But what about `{t, a, c} | {t, b, d}`, should it be merged into `{t, a|b, c|d}`? Probably not because this way, we would lose information. So what would be the criterion? Whether all the elements of the tuples but one are the same?

Reply:
Right, it's not trivial in the generic case.

Maybe tuples that differ in only one element can be a good candidate, but it may be hard to detect them. We need some algorithm to group them. Maybe something similar for unions of records and maps where only one element differs. Let's leave it for another PR.

Reply:
I agree. 🙂

Reply:
> We need some algorithm to group them.

I think we have to do the opposite, i.e. expand them. I've stumbled upon this in practice when dogfooding.

When we have a function with a lot of clauses, a clause that matches on a structure with an inner union needs an assertion from the programmer or an inner case expr to ensure the entire union is handled. This happens even if a previous clause handles a specific case of the structure with one of the union elements.

It's painful because it breaks exhaustiveness checking an requires an seemingly unnecessary annotation.

I think that if we expanded to a union of structures this would be refined and eliminated properly. However, it would come at a cost of having do deal with way bigger unions at the top level.

I'm writing from my phone, I'll follow up with specific examples from the codebase later.

Reply:
Sounds good, but if we expand them, there a risk they will explode. (Too many elements.)

Reply:
Here's the promised example from dogfooding that I think is related to this problem:

https://github.com/josefs/Gradualizer/blob/74338639c0e0b6903aca0f14d5abb73841955b65/src/typechecker.erl#L406-L414

If we do:

```
 compat_ty({type, _, tuple, Args1}, {type, _, tuple, Args2}, Seen, Env) ->
     %% We can assert because we match out `any' in previous clauses.
     %% TODO: it would be perfect if Gradualizer could refine this type automatically in such a case
-    compat_tys(?assert_type(Args1, [type()]),
+    compat_tys(Args1,
                ?assert_type(Args2, [type()]), Seen, Env);
```

we get the following self-check error:

```
ebin/typechecker.beam: The variable on line 413 at column 16 is expected to have type [type()] but it has type any | [abstract_type()]

compat_ty({type, _, tuple, Args1},
          {type, _, tuple, Args2}, Seen, Env) ->
    compat_tys(Args1,
               ^^^^^
```

This happens even though, as the comment says, the `any` element of that union is explicitly matched two clauses above.

Reply:
Moreover, [Lambda Days 2023](https://www.lambdadays.org/lambdadays2023) took place yesterday and today, so I had a chance to talk with [Guillaume Duboc, who works on a type system for Elixir](https://www.irif.fr/users/gduboc/index). He pointed me to [this paper](https://www.irif.fr/~gc/papers/covcon-again.pdf), which suggests using a [disjunctive normal form](https://en.wikipedia.org/wiki/Disjunctive_normal_form), i.e. a union of intersection, to represent types in the subtyping algorithm. This seems to match my intuition about expanding unions for the sake of subtyping and calculating the `type_diff`. I'll try experimenting with the idea.